pkg/validator/cluster.go

```diff
@@ -312,13 +312,61 @@ func validateGeneralClusterChanges(
 	valErrors []string,
 ) []string {
 
-	if cr.Spec.AppID != prevCr.Spec.AppID {
-		appModifiedMsg := fmt.Sprintf(
-			modifiedProperty,
-			"app",
-		)
-		valErrors = append(valErrors, appModifiedMsg)
+	crApp, err := catalog.GetApp(cr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
+	}
+
+	prevCrApp, err := catalog.GetApp(prevCr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
 	}
+
+	if crApp != nil && prevCrApp != nil {
+
+		rbKey := cr.Name + cr.Namespace
+		rbInfo := shared.ClusterRollbackInfoMap[rbKey]
+
+		if prevCrApp.Spec.DistroID != crApp.Spec.DistroID {
+			appModifiedMsg := fmt.Sprintf(
+				invalidDistroId,
+				crApp.Spec.DistroID,
+				prevCrApp.Spec.DistroID,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		if prevCrApp.Spec.Version == crApp.Spec.Version {
+			appModifiedMsg := fmt.Sprintf(
+				versionIsNotModified,
+				crApp.Spec.DistroID,
+				crApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
```

Comment:
Doesn't this trigger if we make some other change to the kdcluster that doesn't touch the app ID?

Also, even if we did change the app ID, is this really an error case or just something we should silently ignore? It would be kind of a weird situation, but it's possible that there are two different app resources with exactly the same distro ID and version. Hmm. Although that really seems like a confusion/problem waiting to happen, and maybe something that our app validator should prevent.

So all that being said, my reaction to this is: this doesn't look like an error situation, it just looks like a situation where we are NOT doing an upgrade or rollback. So we should just skip ahead to any other validation. This is especially true in the case where prevCrApp is the same as crApp. Even if they're different, it's not technically illegal to have two identical kdapp resources, and if we do ever want to prevent that situation from happening then we should enforce that elsewhere.

Reply:
Sure, I changed to comparing `crApp != prevCrApp` (received by clusters `AppId` fields)